<div style="border:solid purple 2px; padding: 20px"> 

**Привет, Виктория!**

Меня зовут Мария Новожилова и я буду делать ревью твоего проекта по предобработке данных. Давай будем общаться на **«ты»**. Если это неприемлемо, обязательно напиши мне в комментариях — мы перейдем на **«вы»**.

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так я выделяю те комментарии, на которые нужно обратить внимание в первую очередь. После их доработки проект будет принят. </div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Зеленым цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Оранжевым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).</div>
    
Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет.

<div class="alert alert-info">
<b>Комментарий студента:</b> Например, такой.</div>   
    
Так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> Привет, новые комментарии отмечены 2V.

<div class="alert alert-success">
<b>Комментарий ревьюера 3V ✔️✔️✔️:</b> Привет, новые комментарии отмечены 3V.

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Хорошая практика начинать проект с описания контекста и задачи. Лучше сделать это описание более подробным.

<div class="alert alert-info">
<b>Комментарий ревьера:</b> Проверка начинается с <a href=#link>шага 3</a></div>   

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<div id='link'></div> 

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Предыдущий раздел заполнен автоматически и безупречен с точки зрения структуры и кода, но надо отметить, что в нём не хватает промежуточных выводов. В следующем задании и в будущих проектах промежуточные и итоговые выводы будут обязательны.

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
debt_from_children = pd.DataFrame()
debt_from_children['count_children'] = data.groupby('children')['debt'].count()
debt_from_children['sum_children'] = data.groupby('children')['debt'].sum()
debt_from_children['result_children'] =\
debt_from_children['sum_children'] / debt_from_children['count_children']

debt_from_children.sort_values('result_children', ascending=False)

count_children  sum_children  result_children
children                                               
4                     41             4         0.097561
2                   2052           194         0.094542
1                   4808           444         0.092346
3                    330            27         0.081818
0                  14091          1063         0.075438
5                      9             0         0.000000

**Вывод:** 
1. Самые ненадежные заемщики с высокой долей просрочек 0,097 - многодетные семьи с 4 детьми. Что обусловлено тем, что с ростом колличества детей в семье увеличиваются ежемесячные расходы.
2. Самые надежные заемщики с низкой долей просрочек 0,075 - заемщики без детей. При этом, отмечается, что семьи с 3 детьми чаще платят в срок, в отличии от семей с 1 ребенком.
3. Исходя из представленных данных у семей имеющих 5 детей отсутствует просрочка по кредитам. Данный вывод скорее всего ложный,(учитывая, что семьи с 4мя детьми имеют просрочку в платежах) является следствием некорректных исходных данных.

<div class="alert alert-danger">
<s><b>Комментарий ревьюера ❌:</b> В целом верно, но важно обратить внимание на то, что некоторые группы слишком малочисленные, их не стоит принимать во внимание. Более корректно сравнивать только многочисленные группы.
    
Также обрати внимание, что print() тут не нужен для вывода, без него будет лучше.
    
И ещё одна рекомендация по коду. Ты верно используешь метод groupby. На самом деле он позволяет получить необходимую таблицу одной строчкой кода. Попробуй оптимизировать свой код.

<div class="alert alert-info">
*Оптимизировала код.*
Пришлось потрудиться, заново разобрать теорию по этой теме и почтитать еще дополнительно про работу groupby и agg, но теперь я поняла эту тему. Код действительно стал короче и читабельнее. Большое спасибо вам, что замотивировали лучше разобраться в теме.

<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> Рада, что мои комментарии оказались полезны! 

Можно упростить код ещё на одну строчку, если добавить в аггрегирующие функции "mean".

<div class="alert alert-info">
Попробовала еще немного скорректировать код по вашему совету. Это просто невероятно! В 2 строчки и готово. Питон увлекает меня все больше и больше.

<div class="alert alert-success">
<b>Комментарий ревьюера 3V ✔️✔️✔️:</b> 👍

In [27]:
debt_from_income = data.groupby('children').agg({'debt':['sum','count', 'mean']}) 

debt_from_income.sort_values('children', ascending=False).head(20)

debt                 
           sum  count      mean
children                       
5            0      9  0.000000
4            4     41  0.097561
3           27    330  0.081818
2          194   2052  0.094542
1          444   4808  0.092346
0         1063  14091  0.075438

In [25]:
debt_from_income = data.groupby('children').agg({'debt':['sum','count']}) 
debt_from_income['conversion'] = debt_from_income['debt']['sum']/debt_from_income['debt']['count']

debt_from_income.sort_values('conversion', ascending=False).head(20)

debt        conversion
           sum  count           
children                        
4            4     41   0.097561
2          194   2052   0.094542
1          444   4808   0.092346
3           27    330   0.081818
0         1063  14091   0.075438
5            0      9   0.000000

<div class="alert alert-info">
*Исправленный вывод:*

1. Самые ненадежные заемщики с высокой долей просрочек семьи с детьми. Что обусловлено тем, что с ростом колличества детей в семье увеличиваются ежемесячные расходы.
2. Самые надежные заемщики с низкой долей просрочек 0,075 - заемщики без детей.
3. Так же отмечается, что семьи с 3 детьми чаще платят в срок, в отличии от семей с 1 ребенком.
4. Исходя из представленных данных у семей имеющих 5 детей отсутствует просрочка по кредитам. Данный вывод скорее всего ложный,(учитывая, что семьи с 4мя детьми имеют просрочку в платежах) является следствием некорректных исходных данных.

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Категории заемщиков с 3-5 детьми слишком малочисленные, поэтому лучше исключить их из выводов.

<div class="alert alert-info">
*Исправленный вывод:*
    
1. Самые ненадежные заемщики с высокой долей просрочек семьи с детьми. Что обусловлено тем, что с ростом колличества детей в семье увеличиваются ежемесячные расходы.
2. Самые надежные заемщики с низкой долей просрочек 0,075 - заемщики без детей.

<div class="alert alert-success">
<b>Комментарий ревьюера 3V ✔️✔️✔️:</b> 👍 корректно

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
debt_from_family_status = pd.DataFrame()
debt_from_family_status['family_status_count'] = data.groupby('family_status')['debt'].count()
debt_from_family_status['family_status_sum'] = data.groupby('family_status')['debt'].sum()
debt_from_family_status['result_family_status'] = \
debt_from_family_status['family_status_sum'] / debt_from_family_status['family_status_count']

debt_from_family_status.sort_values('result_family_status',ascending=False)

family_status_count  family_status_sum  \
family_status                                                   
Не женат / не замужем                 2796                273   
гражданский брак                      4134                385   
женат / замужем                      12261                927   
в разводе                             1189                 84   
вдовец / вдова                         951                 63   

                       result_family_status  
family_status                                
Не женат / не замужем              0.097639  
гражданский брак                   0.093130  
женат / замужем                    0.075606  
в разводе                          0.070648  
вдовец / вдова                     0.066246

**Вывод:** 
Зависимость есть.

1. Люди живущие в гражданском браке и холостые имеют больший процент невозврата кредита в срок.
2. В свою очередь те, кто развелись или овдовели чаще платят в срок, чем люди в браке.

<div class="alert alert-danger">
<s><b>Комментарий ревьюера ❌:</b> Опять же вдовцов маловато для отдельной строчки в выводах о них. Возможно, ты сможешь заметить более крупную закономерность.

<div class="alert alert-info">
*Исправила код.*

In [25]:
debt_from_income = data.groupby('family_status').agg({'debt':['sum','count']}) 
debt_from_income['conversion'] = debt_from_income['debt']['sum']/debt_from_income['debt']['count']

debt_from_income.sort_values('conversion', ascending=False).head(20)

debt        conversion
                       sum  count           
family_status                               
Не женат / не замужем  273   2796   0.097639
гражданский брак       385   4134   0.093130
женат / замужем        927  12261   0.075606
в разводе               84   1189   0.070648
вдовец / вдова          63    951   0.066246

<div class="alert alert-info">
*Исправленный вывод:*
    
1. Люди живущие в гражданском браке и холостые являются ненадежными заемщиками потому, что имеют больший процент невозврата кредита в срок.
2. В свою очередь люди, состоящие в браке, станут более надежными заемщиками так как чаще платят в срок.

<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> Код и выводы стали значительно лучше, молодец!
 
Дополнительно можно заметить такую зависимость: люди, когда-либо состоявшие в официальном браке (в том числе разведенные и вдовые) имеют лучший показатель возврата кредита, чем те, кто не состоял в официальных отношениях.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# В одном из прошлых заданий мы категоризировали уровень дохода создав столбец total_income_category
# попробуем с его помощью получить нужные нам данные

debt_from_income = pd.DataFrame()
debt_from_income['count'] = data.groupby('total_income_category')['debt'].count()
debt_from_income['sum'] = data.groupby('total_income_category')['debt'].sum()
debt_from_income['conversion'] = debt_from_income['sum'] / debt_from_income['count']

debt_from_income.sort_values('conversion', ascending=False).head(20)

count   sum  conversion
total_income_category                         
E                         22     2    0.090909
C                      15921  1353    0.084982
A                         25     2    0.080000
B                       5014   354    0.070602
D                        349    21    0.060172

**Вывод:** 
1. Люди относящиеся к категориям Е и С самые ненадежные заемщики с долей просрочек 0,090 и 0,84.
2. Люди с очень высоким уровнем дохода, относящиеся к категории А, тоже не являются надежными заёмщиками и их доля просрочки составила 0,80. 
3. Самыми надежными заемщиками с низкой долей просрочки оказались люди относящиеся к категории D и В. 

<div class="alert alert-danger">
<s><b>Комментарий ревьюера ❌:</b> Первое, на что мы должны обратить внимание, - несбалансированность групп. Малочисленные группы не дают достоверной статистики и их нужно просто игнорировать в исследовании. Мы можем опираться только на данные групп В и С, как самых многочисленных. 
    
Также в выводах нужно развернуть, какой уровень дохода скрывается за буквами.

<div class="alert alert-info">
*Исправила код.*

In [27]:
debt_from_income = data.groupby('total_income_category').agg({'debt':['sum','count']}) 
debt_from_income['conversion'] = debt_from_income['debt']['sum']/debt_from_income['debt']['count']

debt_from_income.sort_values('conversion', ascending=False).head(20)

debt        conversion
                        sum  count           
total_income_category                        
E                         2     22   0.090909
C                      1353  15921   0.084982
A                         2     25   0.080000
B                       354   5014   0.070602
D                        21    349   0.060172

<div class="alert alert-info">
*Исправленный вывод:*
    
1. Люди, относящиеся к категории С (с уровнем дохода 50001–200000), являются менее надежными заемщиками так как их доля просрочек составляет 0,084.
2. Более надежными заемщиками для банка станут люди из категории В (с уровнем дохода 200001–1000000), для них доля просрочек составила 0,070.

<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> Всё верно, молодец!

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

debt_from_purpose_category = pd.DataFrame()
debt_from_purpose_category['purpose_category_count'] = data.groupby('purpose_category')['debt'].count()
debt_from_purpose_category['purpose_category_sum'] = data.groupby('purpose_category')['debt'].sum()
debt_from_purpose_category['result_purpose_category'] = \
debt_from_purpose_category['purpose_category_sum'] / debt_from_purpose_category['purpose_category_count']

debt_from_purpose_category.sort_values('result_purpose_category', ascending=False)

purpose_category_count  purpose_category_sum  \
purpose_category                                                         
операции с автомобилем                      4279                   400   
получение образования                       3988                   369   
проведение свадьбы                          2313                   183   
операции с недвижимостью                   10751                   780   

                          result_purpose_category  
purpose_category                                   
операции с автомобилем                   0.093480  
получение образования                    0.092528  
проведение свадьбы                       0.079118  
операции с недвижимостью                 0.072551

**Вывод:** 
1. Самые ненадежные заемщики те, что берут кредиты на покупку автомобиля (доля просрочки 0,093) и получение образования (доля просрочки 0,092).
2. Далее по росту надежности следуют заемщики, которые берут кредиты на проведение свадьбы с долей просрочки 0,079.
3. Самые надежные заемщики - те, которые берут кредиты на операции с недвижимостью. У них доля просрочки составляет 0,072.


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> тут всё верно :)

<div class="alert alert-info">
*Исправила код.*

In [29]:
debt_from_income = data.groupby('purpose_category').agg({'debt':['sum','count']}) 
debt_from_income['conversion'] = debt_from_income['debt']['sum']/debt_from_income['debt']['count']

debt_from_income.sort_values('conversion', ascending=False).head(20)

debt        conversion
                          sum  count           
purpose_category                               
операции с автомобилем    400   4279   0.093480
получение образования     369   3988   0.092528
проведение свадьбы        183   2313   0.079118
операции с недвижимостью  780  10751   0.072551

<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> 👍

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Причинами могут быть ошибки ввода данных, человеческий фактор или повреждение данных.

<div class="alert alert-danger">
<s><b>Комментарий ревьюера ❌:</b> Ответ верный и задание действительно можно было понять как общий вопрос, но на самом деле он про данные именно из этого проекта - в каких столбцах в исходных данных к этому проекту больше всего пропусков, какие могут быть причины пропусков конкретно в этом прокте (помимо общих причин)?

<div class="alert alert-info">
Ответ:
В результате исследования было обнаружено 2174 пропуска в столбцах: "days_employed" и "total_income" — общий трудовой стаж в днях и доход. Это пропуски, скорее всего, неслучайные. Люди не работают, соответсвенно не получают доходов. Пропуски можно объяснить тем, что часть клиентов банка работает неофициально, возможно неработающие студенты, так же это могут быть люди в декретном отпуске или неработающие пенсионеры.
Помимо этого необходимо было заменить значения Nan для удобства изучения столбцов.


<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> да, отлично, молодец, что заметила синхронность пропусков

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Когда нам необходимо сохранить массив данных, лучшим решением для колличественных переменных будет заполнить пропуски медианным значением.

<div class="alert alert-danger">
<s><b>Комментарий ревьюера ❌:</b> Не понятно, а как использование медианы влияет на сохранение массива данных?  Пока не вижу корректного ответа на вопрос.

<div class="alert alert-info">
Ответ:
Данные из столбца total_income необходимы для исследования, поэтому важно заполнить пропуски. В столбце income_type данные, по которым мы группируем людей и их доходы. Таким образом считаем медианное значение по каждой категории должности. В данном случае заменить пропуски медианным значением будет лучшим решением, потому что это позволит получить более точные рассчеты, чем если бы мы просто удалили пропуски. При удалении пропусков мы могли получить неправильную картину при рассчете зависимости возврата кредита в срок от уровня доходов.

<div class="alert alert-danger">
<s><b>2V Комментарий ревьюера ❌❌:</b> Виктория, твой ответ верный, но неполный, поэтому возвращаю проект на доработку ещё раз. Ты правильно отметила, что в нашем проекте заполнить пропуски лучше, чем удалить их или оставить незаполненными. Но почему их лучше заполнить медианным значением, а не средним?

<div class="alert alert-info">
Дополнение к ответу:
    В случае с поставленной задачей, заполняем пропуски медианным значением потому, что в столбце total_income некоторые данные сильно выделяются относительно других и, соответственно, среднее значение может дать некорректные результаты.

<div class="alert alert-success">
<b>Комментарий ревьюера 3V ✔️✔️✔️:</b> да, корректно, медиана не чувствительна к выбросам в данных, поэтому даст более корректное значение, чем среднее

### Шаг 4: общий вывод.

Все категории имеют зависимость с выплатой кредита в срок.

1. Наличие детей - чем больше детей, тем выше риск просрочки кредита.
2. Семейное положение - люди живущие в гражданском браке и холостые имеют больший процент невозврата кредита в срок, в отличии от тех, кто развелись или овдовели.
3. Самыми надежными заемщиками с низкой долей просрочки оказались люди относящиеся к категории D с уровнем дохода 30001–50000 и В с уровнем дохода 200001–1000000.
4. Самая надежная цель получения кредита "операции с недвижимостью", доля просрочки составила 0,072, а самые опасные цели получения кредита на покупку автомобиля (доля просрочки 0,093) и получение образования (доля просрочки 0,092).

Таким образом мы выяснили, что все 4 фактора оказывают прямую зависимость к возврату долга в банке. 
Исходя из этого анализа большую зависимость имеют категории Children и family_status.

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>  Общий вывод хороший, молодец. Необходима небольшая корректировка в третьем пункте.

<div class="alert alert-info">
    (Исправила пункт 3)
Все категории имеют зависимость с выплатой кредита в срок.

1. Наличие детей - чем больше детей, тем выше риск просрочки кредита.
2. Семейное положение - люди живущие в гражданском браке и холостые имеют больший процент невозврата кредита в срок, в отличии от тех, кто развелись или овдовели.
3. Надежными заемщиками с низкой долей просрочки оказались люди относящиеся к категории В с уровнем дохода 200001–1000000.
4. Самая надежная цель получения кредита "операции с недвижимостью", доля просрочки составила 0,072, а самые опасные цели получения кредита на покупку автомобиля (доля просрочки 0,093) и получение образования (доля просрочки 0,092).

Таким образом мы выяснили, что все 4 фактора оказывают прямую зависимость к возврату долга в банке. 
Исходя из этого анализа большую зависимость имеют категории Children и family_status.

<div class="alert alert-success">
<b>Комментарий ревьюера 2V ✔️✔️:</b> Выводы стали ещё лучше.

<div style="border:solid purple 2px; padding: 20px"> 
    
**Комментарий ревьюера**

Виктория, не смотря на обилие красных комментариев, твой проект выполнен хорошо и практически готов. Все расчеты верны, их нужно только оптимизировать. Также ты пишешь развернутые грамотные выводы, их опять же нужно скорректировать с учетом несбалансированности групп (малочисленные группы не стоит принимать во внимание). Исправь, пожалуйста, все моменты, отмеченные красными комментариями.
    
Если есть время на доп.задания (желтые комментарии), буду рада их также проверить.
    

Если у тебя есть какие-то вопросы по комментариям или по проекту, напиши их также в своих комментариях, я постараюсь ответить.

<div style="border:solid purple 2px; padding: 20px"> 
    
**Комментарий ревьюера 2V**

Виктория, спасибо за добросовестные доработки, у тебя получился отличный проект. Остался один красный комментарий в последнем вопросе (3.6), поэтому возвращаю проект на доработку ещё раз.

<div style="border:solid purple 2px; padding: 20px"> 
    
**Комментарий ревьюера 3V**

Виктория, теперь всё отлично, с радостью принимаю проект и желаю успехов в дальнейшем обучении.